### ISSUE TITLE
XGBoost does not work for 1.4.0 for hummingbird

### ISSUE THREAD

**Issue by ksaur** [microsoft/hummingbird#495]
In #492 , we needed to pin to `xgboost<1.4.0` because `==1.4.0` ([released](https://github.com/dmlc/xgboost/commit/b9a4f3336af7c78f9f7b66f7c91a419108b7ec83) on April 12th 2021) broke our tests.  We need to investigate why.

From [the pipeline](https://github.com/microsoft/hummingbird/pull/490/checks?check_run_id=2328236386):

```
tests/test_xgboost_converter.py F..FF...sFF...FFF...sFF..            

=================================== FAILURES ===================================
__________ TestXGBoostConverter.test_float64_xgb_classifier_converter __________

self = <test_xgboost_converter.TestXGBoostConverter testMethod=test_float64_xgb_classifier_converter>

    @unittest.skipIf(not xgboost_installed(), reason="XGBoost test requires XGBoost installed")
    def test_float64_xgb_classifier_converter(self):
        warnings.filterwarnings("ignore")
        num_classes = 3
        for max_depth in [1, 3, 8, 10, 12]:
            model = xgb.XGBClassifier(n_estimators=10, max_depth=max_depth)
            np.random.seed(0)
            X = np.random.rand(100, 200)
            y = np.random.randint(num_classes, size=100)
    
            model.fit(X, y)
    
>           torch_model = hummingbird.ml.convert(model, "torch", [])

tests/test_xgboost_converter.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:433: in convert
    return _convert_common(model, backend, test_input, device, extra_config)
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:383: in _convert_common
    return _convert_xgboost(model, backend_formatted, test_input, device, extra_config)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytre...lambda=1, scale_pos_weight=None, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)
backend = 'torch', test_input = [], device = 'cpu'
extra_config = {'container': True, 'n_threads': 2}

    def _convert_xgboost(model, backend, test_input, device, extra_config={}):
        """
        This function is used to generate a *backend* model from a given input [XGBoost] model.
        [XGBoost]: https://xgboost.readthedocs.io/
        """
        assert (
            xgboost_installed()
        ), "To convert XGboost models you need to instal XGBoost (or `pip install hummingbird-ml[extra]`)."
    
        # XGBoostRegressor and Classifier have different APIs for extracting the number of features.
        # In the former case we need to infer them from the test_input.
        if "_features_count" in dir(model):
            extra_config[constants.N_FEATURES] = model._features_count
        elif test_input is not None:
            if type(test_input) is np.ndarray and len(test_input.shape) == 2:
                extra_config[constants.N_FEATURES] = test_input.shape[1]
            else:
                raise RuntimeError(
                    "XGBoost converter is not able to infer the number of input features.\
                        Please pass a test_input to convert or \
>                       fill an issue at https://github.com/microsoft/hummingbird/."
                )
E               RuntimeError: XGBoost converter is not able to infer the number of input features.                    Please pass a test_input to convert or                     fill an issue at https://github.com/microsoft/hummingbird/.

/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:138: RuntimeError
----------------------------- Captured stdout call -----------------------------
[22:13:14] WARNING: ../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
____________ TestXGBoostConverter.test_run_xgb_classifier_converter ____________

self = <test_xgboost_converter.TestXGBoostConverter testMethod=test_run_xgb_classifier_converter>

    @unittest.skipIf(not xgboost_installed(), reason="XGBoost test requires XGBoost installed")
    def test_run_xgb_classifier_converter(self):
        warnings.filterwarnings("ignore")
        for extra_config_param in ["tree_trav", "perf_tree_trav", "gemm"]:
            model = xgb.XGBClassifier(n_estimators=1, max_depth=1)
            np.random.seed(0)
            X = np.random.rand(1, 1)
            X = np.array(X, dtype=np.float32)
            y = np.random.randint(2, size=1)
    
            model.fit(X, y)
    
>           torch_model = hummingbird.ml.convert(model, "torch", [], extra_config={"tree_implementation": extra_config_param})

tests/test_xgboost_converter.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:433: in convert
    return _convert_common(model, backend, test_input, device, extra_config)
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:383: in _convert_common
    return _convert_xgboost(model, backend_formatted, test_input, device, extra_config)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytre...eg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)
backend = 'torch', test_input = [], device = 'cpu'
extra_config = {'container': True, 'n_threads': 2, 'tree_implementation': 'tree_trav'}

    def _convert_xgboost(model, backend, test_input, device, extra_config={}):
        """
        This function is used to generate a *backend* model from a given input [XGBoost] model.
        [XGBoost]: https://xgboost.readthedocs.io/
        """
        assert (
            xgboost_installed()
        ), "To convert XGboost models you need to instal XGBoost (or `pip install hummingbird-ml[extra]`)."
    
        # XGBoostRegressor and Classifier have different APIs for extracting the number of features.
        # In the former case we need to infer them from the test_input.
        if "_features_count" in dir(model):
            extra_config[constants.N_FEATURES] = model._features_count
        elif test_input is not None:
            if type(test_input) is np.ndarray and len(test_input.shape) == 2:
                extra_config[constants.N_FEATURES] = test_input.shape[1]
            else:
                raise RuntimeError(
                    "XGBoost converter is not able to infer the number of input features.\
                        Please pass a test_input to convert or \
>                       fill an issue at https://github.com/microsoft/hummingbird/."
                )
E               RuntimeError: XGBoost converter is not able to infer the number of input features.                    Please pass a test_input to convert or                     fill an issue at https://github.com/microsoft/hummingbird/.

/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/hummingbird/ml/convert.py:138: RuntimeError
----------------------------- Captured stdout call -----------------------------
[22:13:15] WARNING: ../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
__________ TestXGBoostConverter.test_xgb_binary_classifier_converter ___________

self = <test_xgboost_converter.TestXGBoostConverter testMethod=test_xgb_binary_classifier_converter>

    @unittest.skipIf(not xgboost_installed(), reason="XGBoost test requires XGBoost installed")
    def test_xgb_binary_classifier_converter(self):
>       self._run_xgb_classifier_converter(2)

```

**Comment by ksaur** [comment]
Perhaps the issue is [related](https://github.com/dmlc/xgboost/commit/b9a4f3336af7c78f9f7b66f7c91a419108b7ec83) to:
```
* Breaking change: All `data` parameters in prediction functions are renamed to `X`
  for better compliance to sklearn estimator interface guidelines.
* Breaking change: XGBoost used to generate some pseudo feature names with `DMatrix`
  when inputs like `np.ndarray` don't have column names.  The procedure is removed to
  avoid conflict with other inputs. (#6605)
```

**Comment by ksaur** [comment]
Closed with #523
